### PR TITLE
Retry platform-level errors in the isolated process for .NET isolated

### PR DESCRIPTION
**Pull request by davidmrdavid**
Supersedes: https://github.com/Azure/azure-functions-durable-extension/pull/2902
Joint work with @andystaples.

**Problem:** In .NET isolated today, platform-level errors **in the isolated process** like OOMs, Timeouts, and sudden process exits are not automatically retried by DF. Instead, those cause the orchestrator to be marked as failed. This is an unintended behavior change for .NET isolated. This PR aims to bring back robustness against those errors.

Most of the changes related to this exist in the `OutOfProcMiddleware` and `RemoteOrchestrationContext` files.

**"Extra scope":** In trying to add automatic E2E tests to validate these improvements, I had to go down a rabbithole of .NET isolated testing issues in our smoke testing infrastructure today.

Long story short, our docker-based test runner for .NET isolated wasn't testing runtime correctness (it was passed the `-NoValidation` flag) because of intermittent race conditions when publishing .NET isolated apps to Docker. I didn't get far enough into debugging this to root cause the issue (@jviau tells me that upgrading the SDK may automatically solve them) but I did spend enough hours on it to conclude that, for the sake of unblocking this item, it would be practical to abandon the Docker-based testing for .NET isolated and instead use the GitHub actions CI environment directly.

Therefore, this PR changes the definition of the .NET isolated smoke testing GitHub action to instead run the tests directly on the "ubuntu-latest" image. It also introduces a new test runner powershell script that's well-suited for simulating platform-level errors. I hope the comments in the code itself clarify the intent, but please let me know if not.

**Ask for reviewers:** the key shippable changes here are in  `OutOfProcMiddleware` and `RemoteOrchestrationContext` . Please look at those carefully. I've added comments explaining some of the quirky bits.

### CODE REVIEW

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs
Comment (andystaples) [inline test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs:23]: Would suggest giving these replayEvidence files test-specific names to avoid conflicts should multiple retry tests run simultaneously. Even if today they are run sequentially, this may not always be true

**Review by andystaples — APPROVED** [review]
Left minor suggestion, otherwise, looks good

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/host.json
Comment (davidmrdavid) [inline test/SmokeTests/OOProcSmokeTests/DotNetIsolated/host.json:15]: this just helps make sure we retry orchestrator replays fast enough after a platform-level error.

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/host.json
Comment (davidmrdavid) [inline test/SmokeTests/OOProcSmokeTests/DotNetIsolated/host.json:19]: We don't time out sooner because our OOM-based test needs time to run out of memory.

File: src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:177]: I don't love the string-based comparison, but it'll be string-based unless we add `WebJobs.Host.Script` to our dependencies. Open to a discussion here, but please note we have a customer waiting so let's look to be practical. I have no strong feelings, just trying to deploy this fix.

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs:74]: Non-blocking nit: Any method named `TryXXX` is usually expected to return a `bool` value indicating whether it succeeded or failed.

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs:143]: Manually throwing `OutOfMemoryException` is a bit of a red flag. It's really only supposed to be thrown by the runtime. It's partly problematic because there are various pieces of code in .NET and elsewhere that treat OOM and similar "fatal" exceptions differently. In some cases, I think it's supposed to crash the process.

I strongly recommend we throw something else, like `SessionAbortedException`, which is designed for this exact scenario.

File: src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask.xml
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask.xml:478]: Any idea why this is showing up in our public documentation? I thought these APIs were internal?

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs
Comment (cgillum) [inline test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs:38]: I'm not sure I understand this comment, and it's making me question whether I understand what this test is supposed to be doing. I'll ask simply: why would we expect an orchestrator to survive an OOM? It sounds like you're saying there was an OOM but it was handled and we kept running in spite of the OOM. Are you actually saying it *recovered from a worker process crash* (caused by the OOM)? Maybe OOM implies crash, but it's not 100% clear that's what you meant.

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs
Comment (cgillum) [inline test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs:64]: Does this comment need to be updated?

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs
Comment (cgillum) [inline test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs:146]: Rather than replicating the same HTTP function once for each scenario, it might make more sense to have a single HTTP trigger that takes an orchestrator function name as a parameter.

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1
Comment (cgillum) [inline test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1:82]: When is it expected that we would enter this `catch` block?

**Review by cgillum — COMMENTED** [review]
This is a great PR! I have some questions and comments.

**Comment by jviau** [comment]
Summarizing some offline discussion:

1. Handling timeouts, worker crashes, and **host** out of memories is good
2. We should not handle _worker_ out of memory. This is language dependent and each language should handle OOM their own way
    - This should be left for future work. We will need to see if there is an existing mechanism for a language worker to signal an orchestration should be retried. If not, that mechanism will need to be added.
3. I will follow up on the host side to see if there is an existing way to detect worker crashes or other platform-level failures (instead of checking exception type by string). If not, we can file an issue for that ask (and proceed with the string type check for now)

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs:74]: yes, you're right. II chose `Try` to signal somehow that the method may throw, but I know the convention makes this confusing. I'll revert this change.

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs
Comment (davidmrdavid) [reply test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs:146]: I don't feel strongly here, but my personal preference is to specific HTTP endpoints for each orchestrator. When testing locally, it saves me time to just copy-paste the entire URL to trigger the orchestrator. 

But it's no big deal. If you feel strongly, I'll generalize it. Just noting this was a deliberate choice.

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1
Comment (davidmrdavid) [reply test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1:82]: I added comments to the script clarifying this here: https://github.com/Azure/azure-functions-durable-extension/pull/2922/commits/2ff4b79d571e2ec288e5e2da17db92c8370b23a1

Long story short - we expect to enter this catch block if any HTTP request to the host fails. This can happen if the host has not started yet, or if has become unresponsive for whatever reason. Recall that we're inducing isolated worker process crashes, and in my experience that can make the host unresponsive for a bit.

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs
Comment (davidmrdavid) [reply test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs]: Good catch, this is the result of copy-pasting.
Updated: https://github.com/Azure/azure-functions-durable-extension/pull/2922/commits/a70d0146fae6771f6cfd834a374407045043a9ec

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs
Comment (davidmrdavid) [reply test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs]: You're right the wording here was not great.

I do not expect an orchestration invocation to continue running beyond an OOM.

What I mean by survive was:
- orchestrator invocation A runs
- invocation A OOMs
- DF extension abandons the invocation
- DF extension retries the orchestrator invocation (now invocation B, every invocation is unique)
- invocation does not OOM now, and completes

So it "survived"  because it retried.

I'm hoping the new comments here make this clearer: https://github.com/Azure/azure-functions-durable-extension/pull/2922/commits/a70d0146fae6771f6cfd834a374407045043a9ec

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs]: Thanks for the pointer. I'm now throwing `SessionAbortedException`, see: https://github.com/Azure/azure-functions-durable-extension/pull/2922/commits/090f8e729a0c584af3863209154f4ecd9ecd4c9c

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs]: fixed here: https://github.com/Azure/azure-functions-durable-extension/pull/2922/commits/090f8e729a0c584af3863209154f4ecd9ecd4c9c

File: src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask.xml
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask.xml:478]: My guess is that since I added method-level comments to the private method, they're made it to the xml. I think this is probably just to power the intellisense? Not sure, just a guess.

**Comment by davidmrdavid** [comment]
Thanks @jviau. I'll delegate this bit to you then:

>I will follow up on the host side to see if there is an existing way to detect worker crashes or other platform-level failures (instead of checking exception type by string). If not, we can file an issue for that ask (and proceed with the string type check for now)

As for the rest, I'll see what we can do. OOM is the key case we're trying to unblock via this PR (the rest are related "freebies"). @andystaples will start an internal thread about this.

File: src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask.xml
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask.xml:478]: It's not just intellisense, but also the public API docs that we publish, like this: https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.webjobs.extensions.durabletask?view=azure-dotnet. If it's not meant to be published, then we should remove it from this xml file. Might be worth doing a bit of research as when when/why non-public APIs are getting included in this file.

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs
Comment (cgillum) [reply test/SmokeTests/OOProcSmokeTests/DotNetIsolated/FaultyOrchestrators.cs:156]: I don't feel strongly so I'll leave it up to you, but it would enable all the usual benefits of code sharing (easier to add new tests, better consistency b/c only one place to make updates, etc.)

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1
Comment (cgillum) [reply test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1:82]: Got it, I didn't realize that `Invoke-RestMethod` raised an error on non-success status codes.

**Comment by jviau** [comment]
I think all of the points I brought up can also be addressed after this PR, but it is something we should commit to doing ASAP.

1. Adopting a more 'official' way to identify platform issues / worker crashes
2. Implementing/consuming a way for workers to elect to abandon a session (which they can then use for OOM scenarios)

File: src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs
Comment (jviau) [inline src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:174]: Did you verify that `WorkerProcessExitException` is actually available here? Looking at the function host code, I don't see the exception actually _thrown_: https://github.com/Azure/azure-functions-host/blob/1088f24c3ae3a6275f18cbf091fa525c2477be91/src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs#L179-L184

Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:174]: Yes, that's how we stumbled upon this error type in the first place. We found this error by forcing a process exit (`Environment.FailFast`) and placing a breakpoint at this position. 

But yeah I admit that I also don't see the explicit throw here. We can investigate further, or I can show you the runtime behavior through the debugger.

File: src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask.xml
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask.xml]: Seems that adding the method-level comments was the culprit.
I removed them (moved them to the method body), and the xml change was undone: https://github.com/Azure/azure-functions-durable-extension/pull/2922/commits/c92726846c351aeb341c80fd3f0705e6b93653c4

**Review by jviau — APPROVED** [review]

**Pull request merged**
